File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        map<char, int> m;
        int i = 0, j = 0, ans = 0, n = s.size();

        while (j < n) {
            // Update the frequency map for the current character
            m[s[j]]++;

            // Check if there is any duplicate character in the current window
            while (m[s[j]] > 1) {
                m[s[i]]--; // Reduce the frequency of the character at 'i'
                i++;       // Shrink the window from the left
            }

            // Calculate the length of the current substring and update the maximum
            ans = max(ans, j - i + 1);

            // Expand the window by moving 'j' forward
            j++;
        }
        return ans;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        
        HashMap<Character, Integer> mp = new HashMap<>();

        int i=0;
        int j=0;
        char [] arr = new char[s.length()];
        arr = s.toCharArray();
        int length = 0;

        while(j < s.length()){
            if(mp.containsKey(arr[j])){
                while(mp.containsKey(arr[j])){
                    mp.remove(arr[i]);
                    i++;
                }
                    mp.put(arr[j], 1 );

            }
            else{
                mp.put(arr[j], mp.getOrDefault(arr[j], 0)+1 );
            }
            j++;
            length = Math.max(length, mp.size());
        }
        return length;
    }
}